### PR TITLE
feat: add --max-suite-reruns option to cap total reruns across suite

### LINKED ISSUE #298: Add option for total maximum number of retries across the entire test suite
Currently, pytest-rerunfailures allows specifying the maximum number of retries per failed test via the `--reruns` option. However, there is no built-in way to limit the total number of retries across the entire test suite as a whole.

Please consider adding a new option (e.g., `--max-suite-retries`) that would allow users to specify the maximum number of times failed tests can be retried collectively during the suite run, regardless of how many tests fail or pass. This would help limit resource usage and avoid excessive retries in large test suites.

Example:
If `--max-suite-retries=5` is set, and multiple tests fail, retries will stop after a total of 5 retries have been attempted across all failing tests.

### PR DESCRIPTION
**Pull request by Borda**
This pull request introduces a new feature to cap the total number of test reruns across the entire suite using the `--max-suite-retries` option. This allows users to limit overall reruns, regardless of per-test retry settings, to better control resource usage in large or flaky test suites. The implementation includes command-line interface changes, core logic updates, and comprehensive tests.

**New Feature: Suite-wide rerun cap**

* Added a `--max-suite-retries` option to the CLI, allowing users to specify a maximum total number of reruns across the entire test suite. Once this cap is reached, no further reruns are performed, regardless of individual test retry settings. [[1]](diffhunk://#diff-d0cba5f2ab7d1f475a36b55dd29073ff2888a687b1fb14d21f41e03f6db85135R123-R131) [[2]](diffhunk://#diff-ff3c479edefad986d2fe6fe7ead575a46b086e3bbcf0ccc86d85efc4a4c63c79L7-R13) [[3]](diffhunk://#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR223-R237)

**Implementation: Core logic and state management**

* Extended the `StatusDB` and related classes to track and increment a suite-wide rerun counter in a thread-safe manner, supporting both in-memory and socket-based backends. [[1]](diffhunk://#diff-d0cba5f2ab7d1f475a36b55dd29073ff2888a687b1fb14d21f41e03f6db85135R442-R453) [[2]](diffhunk://#diff-d0cba5f2ab7d1f475a36b55dd29073ff2888a687b1fb14d21f41e03f6db85135R538-R542) [[3]](diffhunk://#diff-d0cba5f2ab7d1f475a36b55dd29073ff2888a687b1fb14d21f41e03f6db85135R555-R565) [[4]](diffhunk://#diff-d0cba5f2ab7d1f475a36b55dd29073ff2888a687b1fb14d21f41e03f6db85135R579-R587)
* Updated the test execution protocol to check the suite-wide rerun count before performing a rerun, ensuring the global cap is enforced during test runs.

**Testing: Validation of new behavior**

* Added multiple tests to verify that the suite-wide rerun cap works as intended, including edge cases such as zero cap, cap not reached, and interaction with passing tests.

---

- New CLI option `--max-suite-retries` (int, default None = no limit); once the suite-wide rerun count hits the cap, failing tests are logged as final failures without further retry
- `StatusDB` gains thread-safe `increment_suite_reruns()` / `get_suite_reruns()` via `threading.Lock` for the single-process path
- `ServerStatusDB` overrides use the existing `rerunfailures_db` dict with a `"__suite__"` key and an atomic socket `inc` operation in `run_connection`
- `ClientStatusDB` overrides route increment through the socket to the master
- Five new tests covering: cap enforcement, pass-through when under cap, zero disables all reruns, passing tests don't consume the budget, and standalone option no-op
- CHANGES.rst and README.rst updated

Fixes #298

### CODE REVIEW

File: src/pytest_rerunfailures.py [hunk 1]
```diff
@@ -638,6 +684,14 @@ def pytest_runtest_protocol(item, nextitem):
                 item.ihook.pytest_runtest_logreport(report=report)
             else:
                 # failure detected and reruns not exhausted, since i < reruns
+                max_suite_reruns = item.session.config.option.max_suite_retries
+                if max_suite_reruns is not None:
+                    suite_count = db.increment_suite_reruns()
+                    if suite_count > max_suite_reruns:
+                        # suite-wide limit exhausted — log as final failure
+                        item.ihook.pytest_runtest_logreport(report=report)
+                        continue
```
[the remark this answers is not in the document]
Comment (icemac) [reply src/pytest_rerunfailures.py]: @Borda Please re-review this comment whether it is valid.

File: src/pytest_rerunfailures.py [hunk 2]
```diff
@@ -120,6 +121,15 @@ def pytest_addoption(parser):
...
     )
+    group.addoption(
+        "--max-suite-retries",
+        action="store",
+        dest="max_suite_retries",
+        type=int,
+        default=None,
+        help="Maximum total number of reruns across the entire test suite. "
+        "Once this limit is reached, no further reruns will occur.",
+    )
```
[the remark this answers is not in the document]
Comment (icemac) [reply src/pytest_rerunfailures.py:159]: I'd also prefer to call it `--max-suite-reruns` to make it consistent. (The repository itself also contains `rerun` in its name.

**Review by icemac — CHANGES_REQUESTED** [review]
I think we are a step forward now, there are some requested changes, and please also look at the open Copilot conversations.

**Review by icemac — APPROVED** [review]
LGTM.

**Comment by ron8mcr** [comment]
Hi @icemac, is there a timeline for PyPI release with this feature?

**Comment by icemac** [comment]
Hi @ron8mcr, I just released https://pypi.org/project/pytest-rerunfailures/16.5 including this feature.
